server/src/main/java/com/linecorp/centraldogma/server/auth/AuthProvider.java

```diff
@@ -86,13 +87,25 @@ public interface AuthProvider {
     default HttpService webLoginService() {
         // Redirect to the default page: /link/auth/login -> /web/auth/login
         return (ctx, req) -> {
-            String returnTo = ctx.queryParam("return_to");
-            if (returnTo != null) {
-                returnTo += BUILTIN_WEB_LOGIN_PATH;
-            } else {
-                returnTo = BUILTIN_WEB_LOGIN_PATH;
+            final String returnTo = ctx.queryParam("return_to");
+            String finalRedirectPath = BUILTIN_WEB_LOGIN_PATH;
+
+            if (returnTo != null && !returnTo.trim().isEmpty()) {
+                try {
+                    final URI frontendOrigin = new URI(returnTo);
+                    final String host = frontendOrigin.getHost();
+
+                    // Allow only localhost for security reason.
+                    if ("127.0.0.1".equals(host) || "localhost".equals(host)) {
+                        // e.g. http://localhost:3000 + /web/auth/login -> http://localhost:3000/web/auth/login
```

Comment:
Question) Is `returnTo` only used for local development or testing? `http://localhost:3000/web/auth/login` seems valid when a server is running locally.